File: backend/src/services/shop_service.js
const { pool } = require('../db');

async function listItems() {
  const res = await pool.query(`SELECT id, name, price, image_url, created_at, updated_at FROM shop_items ORDER BY created_at DESC`);
  return res.rows;
}

async function getItem(id) {
  const res = await pool.query(`SELECT id, name, price, image_url, created_at, updated_at FROM shop_items WHERE id = $1`, [id]);
  return res.rows[0] || null;
}

async function createItem({ name, price, image_url }) {
  const res = await pool.query(
    `INSERT INTO shop_items (name, price, image_url) VALUES ($1, $2, $3) RETURNING id, name, price, image_url, created_at, updated_at`,
    [name, price, image_url],
  );
  return res.rows[0];
}

async function updateItem(id, { name, price, image_url }) {
  const res = await pool.query(
    `UPDATE shop_items SET name = COALESCE($2, name), price = COALESCE($3, price), image_url = COALESCE($4, image_url), updated_at = NOW() WHERE id = $1 RETURNING id, name, price, image_url, created_at, updated_at`,
    [id, name, price, image_url],
  );
  return res.rows[0] || null;
}

async function deleteItem(id) {
  await pool.query(`DELETE FROM shop_items WHERE id = $1`, [id]);
  return { deleted: true };
}

async function purchaseItem({ userId, itemId }) {
  const client = await pool.connect();
  try {
    await client.query('BEGIN');

    const userRes = await client.query(`SELECT id, coins, role FROM users WHERE id = $1 FOR UPDATE`, [userId]);
    if (userRes.rowCount === 0) throw new Error('User not found');
    const user = userRes.rows[0];

    const itemRes = await client.query(`SELECT id, price FROM shop_items WHERE id = $1`, [itemId]);
    if (itemRes.rowCount === 0) throw new Error('Item not found');
    const item = itemRes.rows[0];

    if (Number(user.coins) < Number(item.price)) {
      throw new Error('Insufficient coins');
    }

    const newCoins = Number(user.coins) - Number(item.price);
    await client.query(`UPDATE users SET coins = $2, updated_at = NOW() WHERE id = $1`, [userId, newCoins]);

    await client.query(`INSERT INTO shop_purchases (user_id, shop_item_id, price_paid) VALUES ($1, $2, $3)`, [userId, itemId, item.price]);

    await client.query(
      `INSERT INTO notifications (title, message, type, target_group, sender_id, sender_role)
       VALUES ($1, $2, 'shop_purchase', 'admins', $3, $4)`,
      [
        'Shop Purchase',
        `A ${user.role || 'user'} purchased item ${itemId} for ${item.price} coins`,
        userId,
        user.role || 'student',
      ],
    );

    await client.query('COMMIT');

    return { success: true, remaining_coins: newCoins };
  } catch (err) {
    await client.query('ROLLBACK');
    throw err;
  } finally {
    client.release();
  }
}

module.exports = {
  listItems,
  getItem,
  createItem,
  updateItem,
  deleteItem,
  purchaseItem,
};

File: lib/services/shop_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/shop_item.dart';
import 'api_base.dart';

class ShopService extends ApiServiceBase {
  ShopService._();
  static final ShopService instance = ShopService._();

  Future<List<ShopItem>> fetchShopItems() async {
    final rows = await getCachedJsonList(
      '/shop_items?select=id,name,price,image_url,created_at&order=created_at.desc&limit=$defaultPageSize',
      ttl: const Duration(minutes: 10),
      disk: true,
    );
    return rows.map<ShopItem>(ShopItem.fromJson).toList(growable: false);
  }

  Future<ShopItem> createShopItem({
    required String name,
    required int price,
    required String imageUrl,
  }) async {
    final uri = buildUri('/shop_items');
    final response = await http.post(
      uri,
      headers: await buildAuthHeaders(),
      body: jsonEncode({
        'name': name,
        'price': price,
        'image_url': imageUrl,
      }),
    ).timeout(const Duration(seconds: 12));

    if (isSuccess(response)) {
      var decoded = jsonDecode(response.body);
      if (decoded is List && decoded.isNotEmpty) decoded = decoded.first;
      if (decoded is Map<String, dynamic>) {
        return ShopItem.fromJson(decoded);
      }
      throw ApiException('Unexpected create item response format');
    }

    throwApiError(response, 'Failed to create shop item');
  }

  Future<ShopItem> updateShopItem({
    required String itemId,
    String? name,
    int? price,
    String? imageUrl,
  }) async {
    final uri = buildUri('/shop_items?id=eq.$itemId');
    final payload = <String, dynamic>{};
    if (name != null) payload['name'] = name;
    if (price != null) payload['price'] = price;
    if (imageUrl != null) payload['image_url'] = imageUrl;

    final response = await http.patch(
      uri,
      headers: await buildAuthHeaders(),
      body: jsonEncode(payload),
    ).timeout(const Duration(seconds: 12));

    if (isSuccess(response)) {
      var decoded = jsonDecode(response.body);
      if (decoded is List && decoded.isNotEmpty) decoded = decoded.first;
      if (decoded is Map<String, dynamic>) {
        return ShopItem.fromJson(decoded);
      }
      throw ApiException('Unexpected update item response format');
    }

    throwApiError(response, 'Failed to update shop item');
  }

  Future<void> deleteShopItem(String itemId) async {
    final uri = buildUri('/shop_items?id=eq.$itemId');
    final response = await http
        .delete(uri, headers: await buildAuthHeaders())
        .timeout(const Duration(seconds: 12));

    if (isSuccess(response)) return;
    throwApiError(response, 'Failed to delete shop item');
  }

  Future<int> purchaseItem({
    required String userId,
    required String itemId,
    required int itemPrice,
    required int currentCoins,
  }) async {
    final newBalance = currentCoins - itemPrice;
    if (newBalance < 0) {
      throw ApiException('Not enough coins to purchase this item');
    }

    // Update the user's coins directly in the users table
    final uri = buildUri('/users?id=eq.$userId');
    final response = await http.patch(
      uri,
      headers: await buildAuthHeaders(),
      body: jsonEncode({'coins': newBalance}),
    ).timeout(const Duration(seconds: 12));

    if (isSuccess(response)) {
      // Log the purchase to the notifications table (since events table is missing)
      try {
        await http.post(
          buildUri('/notifications'),
          headers: await buildAuthHeaders(),
          body: jsonEncode({
            'sender_id': userId,
            'title': 'Item Purchased',
            'message': itemId,
            'type': 'reward',
            'target_group': 'student',
          }),
        ).timeout(const Duration(seconds: 12));
        await http.post(
          buildUri('/notifications'),
          headers: await buildAuthHeaders(),
          body: jsonEncode({
            'sender_id': userId,
            'title': 'Reward purchased',
            'message': 'Your coin balance was updated for this purchase.',
            'type': 'shop',
            'target_group': 'student',
          }),
        ).timeout(const Duration(seconds: 12));
      } catch (e) {
        print('Failed to log purchase: $e');
      }

      return newBalance;
    }

    throwApiError(response, 'Failed to update user coins during purchase');
  }
}
